x/authz: audit updates

robert-zaremba opened the pull request:
## Description

Suggestions and updates related to x/authz internal audit.

ref: #9114 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

ytbigvinny288 reviewed (approved):
y

amaurym commented on proto/cosmos/authz/v1beta1/event.proto:
```diff
@@ -0,0 +1,35 @@
+  // Module which invokes the authorizaiton grant
+  // TODO: not sure if this is needed. It always has the same value
+  string module = 1;
+  // Msg type for which an autorization is granted
```
```suggestion
  // Msg type URL for which an autorization is granted
```

amaurym commented on proto/cosmos/authz/v1beta1/event.proto:
```diff
@@ -0,0 +1,35 @@
+message EventRevoke {
+  // Module which invokes the authorizaiton revokation
+  string module = 1;
+  // Msg type for which an autorization is revoked
```
```suggestion
  // Msg type URL for which an autorization is revoked
```

amaurym commented on proto/cosmos/authz/v1beta1/event.proto:
```diff
@@ -0,0 +1,35 @@
+  // TODO: not sure if this is needed. It always has the same value
+  string module = 1;
+  // Msg type for which an autorization is granted
+  string msg = 2;
```
Should it be `msg_type_url` to be consistent?

robert-zaremba replied:
good point, yes, I will rename it.

robert-zaremba replied:
although `msg` also sounds good for events.

robert-zaremba replied:
I renamed to `msg_type_url`, but don't have preference. So if anyone else have some please comment.

amaurym commented on x/authz/exported/authorizations.go:
```diff
@@ -15,9 +15,22 @@ type Authorization interface {
 
```
We should change MethodName to maybe `MessageName` or `MsgTypeURL`

ref: https://github.com/cosmos/cosmos-sdk/issues/9114#issuecomment-823982059

robert-zaremba replied:
We can't simply change to `MsgTypeURL` because it's an attribute a the GenericAuthorization structure. I agree that for interface it will make sense to change that, so I will update the GenericAuthorization structure: https://github.com/cosmos/cosmos-sdk/pull/9042/commits/155913e73d94376701420cd61b80ee82e4e9c322#diff-8cf7d1dbb9bfe6d2ff0e5c61ea6daf61ff35f10fd3e226980f06ad19961a7e58

amaurym commented on x/authz/simulation/operations.go:
```diff
@@ -21,16 +21,16 @@ import (
-	TypeMsgExecDelegated       = "/cosmos.authz.v1beta1.Msg/ExecAuthorized"
+	TypeMsgGrantAuthorization  = "/cosmos.authz.v1beta1.Msg/Grant"
+	TypeMsgRevokeAuthorization = "/cosmos.authz.v1beta1.Msg/Revoke"
+	TypeMsgExecDelegated       = "/cosmos.authz.v1beta1.Msg/Exec"
```
These will need to be changed with #9139

robert-zaremba replied:
Do we want to do it now (here), or in #9139?

amaurym replied:
I actually did it in #9139 already.

amaurym reviewed (commented):
Made a first quick scan, nice work 👍 , I love the changes:
- method/message names are shorter in proto files
- merge the 2 grpc queries in one
- better keeper methods

I think there will be some conflicts with #9139

aaronc commented on proto/cosmos/authz/v1beta1/authz.proto:
```diff
@@ -7,21 +7,20 @@ import "gogoproto/gogo.proto";
-  // we need some custom naming here so using `MessageName`
-  string method_name = 1 [(gogoproto.customname) = "MessageName"];
+  // Msg, identified by it's type URL, to grant unrestricted permissions to execute
+  string msg = 1;
```
```suggestion
  string msg_type_url = 1;
```

aaronc replied:
Or... we could alternatively consider this

```suggestion
  google.protobuf.Any msg = 1;
```

robert-zaremba replied:
The first suggestion won't work - see https://github.com/cosmos/cosmos-sdk/pull/9042#discussion_r619134467

amaurym replied:
If we use `google.protobuf.Any msg = 1;`, what should its value hold? asaict, only the type_url is needed, so in order to store minimum stuff (and also not to confuse users), I prefer to keep string.

aaronc replied:
Let's just use `string msg_type_url`

robert-zaremba replied:
I'm using `msg` because it allows to use `MsgTypeURL` as method name in the `Authorization` interface (which, imho is more in Go style and more "straight"). Amaury [suggested](https://github.com/cosmos/cosmos-sdk/pull/9042/files#r622043767) to use `GetMsgTypeURL`  as an alternative. If you have a preference then let me know.

aaronc commented on proto/cosmos/authz/v1beta1/event.proto:
```diff
@@ -0,0 +1,35 @@
+
+// EventGrant is emitted on Msg/Grant
+message EventGrant {
+  // Module which invokes the authorizaiton grant
+  // TODO: not sure if this is needed. It always has the same value
+  string module = 1;
```
```suggestion
```

aaronc commented on proto/cosmos/authz/v1beta1/event.proto:
```diff
@@ -0,0 +1,35 @@
+
+// EventRevoke is emitted on Msg/Revoke
+message EventRevoke {
+  // Module which invokes the authorizaiton revokation
+  string module = 1;
```
```suggestion
```

aaronc commented on proto/cosmos/authz/v1beta1/query.proto:
```diff
@@ -9,43 +9,70 @@ option go_package = "github.com/cosmos/cosmos-sdk/x/authz/types";
+//   // Returns any `Authorization` (or `nil`), with the expiration time, granted to the grantee by the granter for the
+//   // provided msg type.
+//   rpc Authorization(QueryAuthorizationRequest) returns (QueryAuthorizationResponse) {
+//     option (google.api.http).get = "/cosmos/authz/v1beta1/granters/{granter}/grantees/{grantee}/grant";
+//   }
+
+//   // Returns list of `Authorization`, granted to the grantee by the granter.
+//   rpc Authorizations(QueryAuthorizationsRequest) returns (QueryAuthorizationsResponse) {
+//     option (google.api.http).get = "/cosmos/authz/v1beta1/granters/{granter}/grantees/{grantee}/grants";
+//   }
+// }
+
+// // QueryAuthorizationRequest is the request type for the Query/Authorization RPC method.
+// message QueryAuthorizationRequest {
+//   string granter     = 1;
+//   string grantee     = 2;
+//   string method_name = 3;
+// }
+
+// // QueryAuthorizationResponse is the response type for the Query/Authorization RPC method.
+// message QueryAuthorizationResponse {
+//   // authorization is a authorization granted for grantee by granter.
+//   cosmos.authz.v1beta1.AuthorizationGrant authorization = 1;
+// }
+
+// // QueryAuthorizationsRequest is the request type for the Query/Authorizations RPC method.
+// message QueryAuthorizationsRequest {
+//   string granter = 1;
+//   string grantee = 2;
+//   // pagination defines an pagination for the request.
+//   cosmos.base.query.v1beta1.PageRequest pagination = 3;
+// }
+
+// // QueryAuthorizationsResponse is the response type for the Query/Authorizations RPC method.
+// message QueryAuthorizationsResponse {
+//   // authorizations is a list of grants granted for grantee by granter.
+//   repeated cosmos.authz.v1beta1.AuthorizationGrant authorizations = 1;
+//   // pagination defines an pagination for the response.
+//   cosmos.base.query.v1beta1.PageResponse pagination = 2;
+// }
```
```suggestion
```

aaronc commented on proto/cosmos/authz/v1beta1/tx.proto:
```diff
@@ -6,57 +6,59 @@ import "gogoproto/gogo.proto";
+// MsgGrantRequest grants the provided authorization to the grantee on the granter's
 // account with the provided expiration time.
-message MsgGrantAuthorizationRequest {
+message MsgGrantRequest {
```
```suggestion
message MsgGrant {
```

aaronc commented on proto/cosmos/authz/v1beta1/tx.proto:
```diff
@@ -6,57 +6,59 @@ import "gogoproto/gogo.proto";
 // authorizations granted to the grantee. Each message should have only
 // one signer corresponding to the granter of the authorization.
-message MsgExecAuthorizedRequest {
+message MsgExecRequest {
```
```suggestion
message MsgExec {
```

aaronc commented on proto/cosmos/authz/v1beta1/tx.proto:
```diff
@@ -6,57 +6,59 @@ import "gogoproto/gogo.proto";
+// MsgRevokeRequest revokes any authorization with the provided sdk.Msg type on the
 // granter's account with that has been granted to the grantee.
-message MsgRevokeAuthorizationRequest {
+message MsgRevokeRequest {
```
```suggestion
message MsgRevoke {
```

robert-zaremba replied:
I thought we want to have `Request` suffix for all request argument types.

aaronc commented on x/authz/exported/authorizations.go:
```diff
@@ -10,14 +10,27 @@ import (
-	// MethodName returns the fully-qualified Msg service method name as described in ADR 031.
-	MethodName() string
+	// MsgTypeURL returns the fully-qualified Msg service method name as described in ADR 031.
+	MsgTypeURL() string
```
I wonder if we could just do something like this:

```suggestion
	MsgType() sdk.Msg
```

What do you think?

robert-zaremba replied:
So to remove `URL` everywhere? (in all `msg_type_url` names)

aaronc replied:
No. SendAuthorization already knows it's concrete type url for example. Just an idea...

robert-zaremba replied:
We decide to keep the `MsgTypeURL() string` - it's easier to reason. We don't need an instance here. Only a type name for routing.

aaronc commented on x/authz/exported/authorizations.go:
```diff
@@ -10,14 +10,27 @@ import (
 type Authorization interface {
```
Can we move this out of the `exported` package?

robert-zaremba replied:
yes, we should. I forgot to put it in the task list.

robert-zaremba commented:
I've moved all Genesis functions to Keeper, and moved exported to the module root package. So we don't have the weird `exported.XXX` and instead we use a clear name: `authz.XXX`. 
BTW: does any one have a suggestion to rename `authz.Authorization`? Or shall we live it?

aaronc commented:
> BTW: does any one have a suggestion to rename `authz.Authorization`? Or shall we live it?

I think it's probably fine...

aaronc commented on simapp/app.go:
```diff
@@ -82,8 +82,8 @@ import (
 
-	authz "github.com/cosmos/cosmos-sdk/x/authz"
 	authzkeeper "github.com/cosmos/cosmos-sdk/x/authz/keeper"
+	authz_m "github.com/cosmos/cosmos-sdk/x/authz/module"
```
Why not just `authz`?

robert-zaremba replied:
Because the `authz` package is `github.com/cosmos/cosmos-sdk/x/authz`.

amaurym commented on proto/cosmos/authz/v1beta1/event.proto:
```diff
@@ -0,0 +1,30 @@
+syntax = "proto3";
+package cosmos.authz.v1beta1;
+
+// import "cosmos_proto/cosmos.proto";
+// import "google/protobuf/timestamp.proto";
+// import "gogoproto/gogo.proto";
+// import "google/protobuf/any.proto";
+
```
```suggestion
```

amaurym commented on x/authz/authorizations.go:
```diff
@@ -0,0 +1,38 @@
+
+	// MsgTypeURL returns the fully-qualified Msg service method URL (as described in ADR 031),
+	// which will process and accept or reject a request.
+	MsgTypeURL() string
```
@robert-zaremba You mentioned this might conflict with https://github.com/cosmos/cosmos-sdk/pull/9042#discussion_r621311724, right?

Maybe an idea here is `GetMsgTypeURL`?

robert-zaremba replied:
In `GenericAuthorization` I have `msg` - I like the short name.  But if you prefer, I can change it and use `Get...` here.

robert-zaremba replied:
Also, `Get...` is not very go style.

amaurym commented on x/authz/authorizations.go:
```diff
@@ -0,0 +1,38 @@
+	// which will process and accept or reject a request.
+	MsgTypeURL() string
+
+	// Accept determines whether this grant permits the provided sdk.ServiceMsg to be performed, and if
+	// so provides an upgraded authorization instance.
+	Accept(ctx sdk.Context, msg sdk.ServiceMsg) (AcceptResponse, error)
```
I really would prefer to merged https://github.com/cosmos/cosmos-sdk/pull/9139 first, is that okay? Because this needs to be updated to remove `sdk.ServiceMsg`

robert-zaremba replied:
OK, I started the review few days ago, but didn't finish.

amaurym commented on x/staking/types/authz.go:
```diff
@@ -25,43 +25,44 @@ func NewStakeAuthorization(allowed []sdk.ValAddress, denied []sdk.ValAddress, au
 
 // Accept implements Authorization.Accept.
-func (authorization StakeAuthorization) Accept(ctx sdk.Context, msg sdk.ServiceMsg) (updated authz.Authorization, delete bool, err error) {
+// func (authorization StakeAuthorization) Accept(ctx sdk.Context, msg sdk.ServiceMsg) (updated authz.Authorization, delete bool, err error) {
```
```suggestion
```

amaurym reviewed (changes requested):
lgtm overall.

Blocking because i'd like to see #9139 merged first, and then we should do another pass here to make sure we're not using ServiceMsgs anymore.

If it's in scope for this PR, i'd to see `authz/types` stuff moved into authz root.

robert-zaremba commented:
> i'd love to see authz/types stuff moved into authz root.

I've just did it :)

aaronc commented on x/authz/codec/codec.go:
```diff
@@ -4,26 +4,26 @@ import (
-		(*exported.Authorization)(nil),
+		(*authz.Authorization)(nil),
 		&bank.SendAuthorization{},
-		&GenericAuthorization{},
+		&authz.GenericAuthorization{},
 		&staking.StakeAuthorization{},
```
Woahh... we should not be registering `bank` and `staking` authorizations here. These should be registered by the bank and staking modules. Then we can move this out of it's own package and back into `x/authz`

robert-zaremba replied:
totally, I will move it.

robert-zaremba commented:
ADR-31 / master merged and conflict resolved. Ready for final review.

amaurym commented on x/authz/keeper/keys.go:
```diff
@@ -0,0 +1,29 @@
+
+// addressesFromGrantStoreKey - split granter & grantee address from the authorization key
+func addressesFromGrantStoreKey(key []byte) (granterAddr, granteeAddr sdk.AccAddress) {
+	// key if of format:
```
```suggestion
	// key is of format:
```

amaurym commented on x/authz/keys.go:
```diff
@@ -0,0 +1,25 @@
+	QuerierRoute = ModuleName
+)
+
+// Keys for authz store
+// Items are stored with the following key: values
+//
+// - 0x01<granterAddressLen (1 Byte)><granterAddress_Bytes><granteeAddressLen (1 Byte)><granteeAddress_Bytes><msgType_Bytes>: Grant
+
+var (
+	// Keys for store prefixes
+	GrantKey = []byte{0x01} // prefix for each key
+)
```
All this should imo go into keeper/keys.go

robert-zaremba replied:
I was thinking about it - but only the store keys (used by keeper). This is how all other modules are organizing the keys. But I agree - it's better to keep keeper keys private to the keeper. If you agree I will move the store keys to keeper.

amaurym replied:
Yes, I was talking about the store keys only (the highlighted part in the comment). +1 for moving to keeper.

amaurym commented on x/authz/simulation/operations.go:
```diff
@@ -23,21 +23,21 @@ import (
 	// TODO Remove `Request` suffix
 	// https://github.com/cosmos/cosmos-sdk/issues/9114
-	TypeMsgGrantAuthorization  = sdk.MsgTypeURL(&types.MsgGrantAuthorizationRequest{})
-	TypeMsgRevokeAuthorization = sdk.MsgTypeURL(&types.MsgRevokeAuthorizationRequest{})
-	TypeMsgExecDelegated       = sdk.MsgTypeURL(&types.MsgExecAuthorizedRequest{})
+	TypeMsgGrantAuthorization  = sdk.MsgTypeURL(&authz.MsgGrant{})
```
You can remove the comment above too

amaurym commented on x/authz/spec/01_concepts.md:
```diff
@@ -4,8 +4,14 @@ order: 1
+Authorizations use the new `Msg` type from [ADR 031](../../../architecture/adr-031-msg-service.md).
 
 
 +++ https://github.com/cosmos/cosmos-sdk/blob/c95de9c4177442dee4c69d96917efc955b5d19d9/x/authz/types/authorizations.go#L15-L24
```
Snippet should be updated

robert-zaremba replied:
Yes, I created a task to update this after the merge: https://github.com/cosmos/cosmos-sdk/issues/9257. Thanks.

amaurym commented on x/authz/spec/01_concepts.md:
```diff
@@ -4,8 +4,14 @@ order: 1
+Grant is an allowance to execute an Msg by grantee address on behalf of the granter.
+Authorization is an interface which must be implemented by a concrete authorization logic to validate and execute grants. They are extensible and can be defined for any Msg service method even outside of the module where the Msg method is defined. See the `SendAuthorization` example below for more details.
+
+Authorizations use the new `Msg` type from [ADR 031](../../../architecture/adr-031-msg-service.md).
```
This is not really true anymore, the Msg TypeURL was introduced in adr-020, adr-031 only talks about Msg service

robert-zaremba replied:
Yes, I was updating the spec few weeks ago.

robert-zaremba replied:
I removed the line.

amaurym commented on x/authz/spec/01_concepts.md:
```diff
@@ -17,7 +23,7 @@ Cosmos-SDK `x/authz` module comes with following authorization types
 ### SendAuthorization
 
-`SendAuthorization` implements `Authorization` interface for the `cosmos.bank.v1beta1.Msg/Send` ServiceMsg, that takes a `SpendLimit` and updates it down to zero.
+`SendAuthorization` implements `Authorization` interface for the `cosmos.bank.v1beta1.Msg/Send` Msg, that takes a `SpendLimit` and updates it down to zero.
```
```suggestion
`SendAuthorization` implements `Authorization` interface for the `cosmos.bank.v1beta1.MsgSend` Msg, that takes a `SpendLimit` and updates it down to zero.
```

amaurym commented on x/authz/spec/01_concepts.md:
```diff
@@ -28,13 +34,14 @@ Cosmos-SDK `x/authz` module comes with following authorization types
 +++ https://github.com/cosmos/cosmos-sdk/blob/c95de9c4177442dee4c69d96917efc955b5d19d9/x/authz/types/generic_authorization.go#L20-L28
 
-- `method_name` holds ServiceMsg type.
+- `method_name` stores Msg type URL.
```
```suggestion
- `msg` stores Msg type URL.
```

amaurym commented on x/authz/spec/02_state.md:
```diff
@@ -6,7 +6,7 @@ order: 2
 ## AuthorizationGrant
 
-Authorizations are identified by combining granter address (the address bytes of the granter), grantee address (the address bytes of the grantee) and ServiceMsg type (its method name).
+Grants are identified by combining granter address (the address bytes of the granter), grantee address (the address bytes of the grantee) and Msg type (its URL). Hence we only allow one grant for the (granter, grantee, msg type) triple.
```
```suggestion
Grants are identified by combining granter address (the address bytes of the granter), grantee address (the address bytes of the grantee) and Msg type (its TypeURL). Hence we only allow one grant for the (granter, grantee, msg type) triple.
```

amaurym commented on x/authz/spec/04_events.md:
```diff
@@ -4,25 +4,4 @@ order: 4
-| revoke-authorization | grant-type        | {msgType}          |
-| revoke-authorization | granter           | {granterAddress}   |
-| revoke-authorization | grantee           | {granteeAddress}   |
+The authz module emits proto eevents defined in [](../../../docs/core/proto-docs.md#cosmos/authz/v1beta1/event.proto)
```
```suggestion
The authz module emits proto events defined in [the Protobuf reference](../../../docs/core/proto-docs.md#cosmos/authz/v1beta1/event.proto).
```

amaurym commented on x/staking/types/authz.go:
```diff
@@ -3,61 +3,64 @@ package types
 // Tracking issues https://github.com/cosmos/cosmos-sdk/issues/9054, https://github.com/cosmos/cosmos-sdk/discussions/9072
 const gasCostPerIteration = uint64(10)
 
+// Normalized Msg type URLs
 var (
 	_ authz.Authorization = &StakeAuthorization{}
 )
 
+var _ authz.Authorization = &StakeAuthorization{}
+
```
These changes seem weird

robert-zaremba replied:
why it seams weird? I'm not sure, probably `lint-fix` updated it.

amaurym replied:
It's duplicated code, and the comment doesn't make sense.

robert-zaremba replied:
ah, I is see. will remove the line.

robert-zaremba commented on proto/cosmos/authz/v1beta1/authz.proto:
```diff
@@ -6,22 +6,21 @@ import "google/protobuf/timestamp.proto";
   option (cosmos_proto.implements_interface) = "Authorization";
 
-  // method name to grant unrestricted permissions to execute
-  // Note: MethodName() is already a method on `GenericAuthorization` type,
-  // we need some custom naming here so using `MessageName`
-  string method_name = 1 [(gogoproto.customname) = "MessageName"];
+  // Msg, identified by it's type URL, to grant unrestricted permissions to execute
+  string msg = 1;
 }
 
-// AuthorizationGrant gives permissions to execute
+// Grant gives permissions to execute
 // the provide method with expiration time.
-message AuthorizationGrant {
+message Grant {
   google.protobuf.Any       authorization = 1 [(cosmos_proto.accepts_interface) = "Authorization"];
   google.protobuf.Timestamp expiration    = 2 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
 }
```
But then we will need to move `expiration` to all `Authorizaiton` concrete types. It's possible - but to keep that requirement, we will probably need to extend the `Authorization` interface by `ExpiresAt() time.Time`.

amaurym replied:
As per a Discord discussion, here are the pros and cons I can think of (ordered by importance, according to me).

**Pros of changing to 1 proto message**
  - Simplicity.
    - 1 grant type == 1 proto message, no nested stuff.
  - Centralized logic.
    - Everything happens in `Accept()`. There's no other magic (say pruning based on expiration date) that the user might need to read docs/code to know.
  - Possibility to have grants without expiration date (semantically).
  - State size is smaller.

**Cons of changing to 1 proto message**
  - We want to release `beta` asap. This change might need another ADR round.
  - No way to add a secondary index for expiration, which would allow pruning based on expiration.
  - Maybe slightly more code for grant creators: expiration logic is duplicated in each grant (though can be easily factorized into a function).

aaronc replied:
> Centralized authorization logic.
> Everything happens in Accept(). There's no other magic (say pruning based on expiration date) that the user might need to read docs/code to know.
> Possibility to have grants without expiration date (without a expiration = nil hack).

These seem like cons to me.

I would also add:

**Cons:**
* client developers can't write user interfaces that show `expiration` as a top-level field for authorizations
* every `Authorization` needs to repeat `expiration` logic
* some `Authorization` types might accidentally be missing `expiration` as an option

aaronc replied:
In general, I don't see a lot of value in moving something which can be generalized up into a polymorphic `Any` type, especially with protobuf. We do this in `x/authz` and now clients need to switch on every `Account` interface type just to get the account number and sequence number which all `Account`s must have.

amaurym replied:
OK. Let's keep the current way then.

robert-zaremba commented on x/authz/keeper/keeper.go:
```diff
@@ -34,30 +33,30 @@ func NewKeeper(storeKey sdk.StoreKey, cdc codec.BinaryCodec, router *baseapp.Msg
-// getAuthorizationGrant returns grant between granter and grantee for the given msg type
-func (k Keeper) getAuthorizationGrant(ctx sdk.Context, grantStoreKey []byte) (grant types.AuthorizationGrant, found bool) {
+// getGrant returns grant stored at skey.
+func (k Keeper) getGrant(ctx sdk.Context, skey []byte) (grant authz.Grant, found bool) {
```
maybe just `key`? It's a private function and it's documented.

amaurym replied:
OK I just read the docs, let's leave `skey`.

amaurym commented on x/authz/keeper/grpc_query.go:
```diff
@@ -13,39 +13,56 @@ import (
-	var authorizations []*types.AuthorizationGrant
+
+	if req.MsgTypeUrl != "" {
+		authorization, expiration := k.GetCleanAuthorization(ctx, grantee, granter, req.MsgTypeUrl)
```
Could we add a comment that since this is in a query, `ctx` is ephemeral so `k.GetCleanAuthorization()` doesn't actually delete the authorization from state

robert-zaremba replied:
it will depend on who is calling the method. If it's called in a Tx context (other msg in a tx context calling this method), then it may change the state.

amaurym reviewed (approved):
lgtm, pre-approving pending conflicts

robert-zaremba commented:
I prefer to fix conflicts once after having second approval. Let's wrap it and close it.

aaronc reviewed (approved):
Still a couple things to clean up but pre-approving

blushi reviewed (commented):
Should we also add some changelog to the authz ADR and update it according to these changes?

robert-zaremba commented:
> Should we also add some changelog to the authz ADR and update it according to these changes?

Authz is new in 0.43, so we shouldn't put changes for not released things in a Changelog.

robert-zaremba commented on x/authz/codec/codec.go:
```diff
@@ -0,0 +1,25 @@
+package types
```
yes, moved.

robert-zaremba commented on x/authz/client/cli/query.go:
```diff
@@ -27,113 +26,68 @@ func GetQueryCmd() *cobra.Command {
+			fmt.Sprintf(`Query authorization grants for a granter-grantee pair. If msg-type-url
+is set, it will select grants only for that msg type.
+Examples:
+$ %s query %s grants cosmos1skj.. cosmos1skjwj..
+$ %s query %s grants cosmos1skjw.. cosmos1skjwj.. %s
```
updated CLI cmd info

robert-zaremba commented:
> Should we also add some changelog to the authz ADR and update it according to these changes?

Yes, doing it in a separate PR: https://github.com/cosmos/cosmos-sdk/pull/9270

blushi commented:
> > Should we also add some changelog to the authz ADR and update it according to these changes?
> 
> Authz is new in 0.43, so we shouldn't put changes for not released things in a Changelog.

I meant the changelog at the beginning of the ADR (not the general changelog file)

aaronc commented:
There are failing simulations. Can you check this @robert-zaremba ?

alessio commented:
This should be easy I guess:

```
# github.com/cosmos/cosmos-sdk/simapp [github.com/cosmos/cosmos-sdk/simapp.test]
simapp/sim_test.go:177:13: undefined: "github.com/cosmos/cosmos-sdk/x/authz".StoreKey
```

technicallyty commented:
looks like this [comment](https://github.com/cosmos/cosmos-sdk/commit/59810f3286420ab73452b635aebeffac9146e355#diff-be19bed5ed22bc822a94ceca6765ad702eda2344f639af6f50a01408ed543abfR20-R21) wasn't updated to reflect one of the changes in this PR

robert-zaremba commented:
Thanks @technicallyty  - I will update the comment int https://github.com/cosmos/cosmos-sdk/pull/9270
